Issue69 bulk update links

coleshaw opened the pull request:
Close #69 
Close #70 

@graft , what are your thoughts on this `bulk_update` route? It seems kind of busy / messy, since I basically have to loop through all the revisions twice. Can probably tighten it up, along with the existing routes, but wanted to get some feedback on the overall approach before refactoring.

graft commented:
I think in general it's the right approach. However I'm a little concerned about the litany of tests which seems pretty non-systematic. What I suspect is that we need a low-level, safe operations interface that gets used by all controllers and would apply here. This seems a large project to accomplish at once.

I am also thinking about how the order of operations matters in determining what is valid, which seems very hard to satisfy.

coleshaw commented:
Yeah, neither the code nor the tests are very DRY -- I basically copied the code and tests from both `remove` and `copy` to make sure that the `bulk_update` route covered at least the same basic use cases.

Do you want to me to at least try to refactor the `file_controller` routes to have more shared code / tests? Or save that for a future lower-level interface?

For order of validity checks, my approach was to reject as soon as possible ... but after the low hanging fruit, not sure how to best order the checks.

coleshaw commented:
Just a couple thoughts on a low-level interface....

I feel like there are two types of checks at the controller level. One is the "syntax" check, like are all the parameters there, are they in the right format, etc. The other is the "authorization" check, like can this user view the bucket, edit the file, delete the file, etc. Is the authorization check the one that you're thinking can go into a low-level interface? Feels like something we could bake into the models (Metis::File.can_user_edit?) or have an internal service provider that does the checks for you (Metis::Authorization.can_user_edit_file?)...or a wrapper around each object that handles the checks (Metis::FileWithAuthz(File))? Okay, certainly some random thoughts.

graft commented:
Let's cut scope here. We can leave out deletion since we agree it is not
strictly necessary, it can be added in later. That leaves a bulk copy
operation. Since everything is going into a single magma project bucket all
we care about is that the source files are legible - they exist and the
user can read them. This seems much easier to manage without a low-level
operations interface, which we can design later.

On Wed, May 20, 2020, 6:52 AM coleshaw <notifications@github.com> wrote:

> Just a couple thoughts on a low-level interface....
>
> I feel like there are two types of checks at the controller level. One is
> the "syntax" check, like are all the parameters there, are they in the
> right format, etc. The other is the "authorization" check, like can this
> user view the bucket, edit the file, delete the file, etc. Is the
> authorization check the one that you're thinking can go into a low-level
> interface? Feels like something we could bake into the models
> (Metis::File.can_user_edit?) or have an internal service provider that does
> the checks for you (Metis::Authorization.can_user_edit_file?)...or a
> wrapper around each object that handles the checks
> (Metis::FileWithAuthz(File))? Okay, certainly some random thoughts.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/mountetna/metis/pull/73#issuecomment-631487400>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AACHPAZ4U6YGLLTHXCRQSRTRSPOCXANCNFSM4NFJSWHA>
> .
>

coleshaw commented:
@graft , thoughts on this? I've left space for the remove functionality and removed related "remove" tests. Still not super clean, I think because of validating everything first...but probably won't be super clean until we get that low level interface...

graft commented on lib/server/controllers/file_controller.rb:
```diff
@@ -104,4 +107,91 @@ def copy
+        if Metis::File.exists?(new_file_name, new_bucket, new_folder)
+          old_dest_file = get_file_obj_from_path(revision[:dest])
+          old_dest_file.remove!
+        end
```
Why not just update the datablock instead of removing/recreating the file?

graft commented on lib/server/controllers/file_controller.rb:
```diff
@@ -104,4 +107,91 @@ def copy
+      map {|rev| extract_bucket_from_path(rev[:source])}.uniq
+    all_dest_buckets = revisions.select {|rev| rev[:dest]}.
+      map {|rev| extract_bucket_from_path(rev[:dest])}.uniq
+
```
This might be nice in a revision object that collects all of these validations and bucket-extractions and so on.

graft commented on lib/server/controllers/file_controller.rb:
```diff
@@ -104,4 +107,91 @@ def copy
+    user_authorized_bucket_names = Metis::Bucket.where(
+      project_name: @params[:project_name]
+    ).all.select{|b| b.allowed?(@user, @request.env['etna.hmac'])}.
+      map {|bucket| bucket.name}
```
It might be easier to break down `Metis::Bucket#allowed?` here, and just extract the `id` (if any) from the valid hmac. Then you could retrieve all buckets `.where(owner: [ 'metis', id ], name: all_source_buckets + all_dest_buckets)`.

coleshaw replied:
Would that do the same ROLES check, if I skip `allowed?`? Seems like this would be actually more restrictive, right, only letting users copy to / from buckets where they are owners?

graft replied:
You're right, it wouldn't run those ROLES checks, and doing those checks in a db call is hard, so you'd have to use `allowed?` anyway, but adding in the owner and name requirements might let you pare down the number of buckets you are checking.

coleshaw commented:
@graft , I've added in a `Revision` class and also cleaned up the controller code. I left the tests mostly as-is, was kind of torn between more low-level unit tests for the new shared methods, or  leaving the higher-level functional tests (which might be clearer in terms of expected behavior of the controllers?).

graft commented on lib/revision.rb:
```diff
@@ -0,0 +1,106 @@
+    end
+
+    def self.extract_bucket_name_from_path(path)
+      # Assumes Metis file paths are in the form
```
Not to go too deep down the rabbit hole of abstraction here, but since we are committing to this "metis path" notion, it probably deserves its own service object, e.g. `MetisPath.new(path).bucket_name`

graft replied:
A lot of the methods before could go away if we had `@source = MetisPath.new(source_path)`.

graft commented on lib/revision.rb:
```diff
@@ -0,0 +1,106 @@
+      raise Etna::BadRequest, 'All revisions require "source" parameter' unless params.has_key? :source
+      raise Etna::BadRequest, 'All revisions require "dest" parameter' unless params.has_key? :dest
+
+      raise Etna::BadRequest, 'Invalid source path' unless valid_file_path?(params[:source])
```
Since we can't be sure we're in a request here, we might not want to raise BadRequest to get out of here. Another option might be to just save the params and save these validations for a `valid?` method, which the controller can use to check if the request is valid and then raise BadRequest, etc. If not that we might raise out of here with some other error that we catch in the controller, but since this object has nothing better to do than provide validations it seems fine to leave them as methods rather than raising.

coleshaw replied:
I didn't pass in the request -- thoughts on this `valid?` method? For bulk copy, I was trying to avoid fetching bucket for each revision, so passing in the list of authorized bucket names directly.

graft commented on lib/server/controllers/file_controller.rb:
```diff
@@ -72,36 +74,115 @@ def rename
     bucket = require_bucket
 
-    file = Metis::File.from_path(bucket, @params[:file_path])
+    raise Etna::Forbidden, "Cannot edit project #{@params[:project_name]}" unless @user.can_edit?(@params[:project_name])
```
Hmm, should the route just be `can_edit? :project_name` instead of this? I think I initially thought: they only need to be able to read the file in the route path, not edit it - the edit permissions are for the write destination. If they are in the same project then the `can_edit?` permission should cover this. If not, the route should be `can_view?` and we should check if they can edit `new_project_name`. Since we don't support the latter at the moment it's probably easiest to just make the route `can_edit?`

coleshaw replied:
Oh yeah, I was wondering about this for the source file versus the destination file. Makes sense to just add it to the route definition, since we don't support changing projects at the moment.

graft commented on lib/server/controllers/file_controller.rb:
```diff
@@ -72,36 +74,115 @@ def rename
+    JSON.parse(@params[:revisions]).
+      each { |rev|
+        revisions.push(Metis::CopyRevision.new(rev.transform_keys(&:to_sym)))
+      }
```
can do like: `JSON.parse(@params[:revisions], symbolize_names: true).map{|r| Metis::CopyRevision.new(rev)}`

coleshaw commented:
@graft refactored per your feedback. Thoughts? If it looks good I'll continue future work on different branches, to prevent this PR from getting too big.

graft commented on lib/models/file.rb:
```diff
@@ -105,6 +105,35 @@ def self.exists?(file_name, bucket, parent_folder)
+      dest_folder_path, dest_file_name = Metis::File.path_parts(params[:dest_file_path])
+
+      dest_bucket = Metis::Bucket.where(
+        project_name: params[:project_name],
```
Can also be: `Metis::Bucket.find()`

graft commented on lib/path.rb:
```diff
@@ -0,0 +1,66 @@
+      end
+
+      def self.extract_bucket_name_from_path(path)
+        return FILEPATH_MATCH.match(path)[:bucket_name]
```
Does this have value as a class method? Could these just be folded into `#bucket_name`, etc?

coleshaw replied:
Only have it as a class method because I was using it in the `metis_controller` methods. Can remove it after I refactor those.

graft commented on lib/server/controllers/file_controller.rb:
```diff
@@ -72,36 +74,116 @@ def rename
-      data_block: file.data_block
-    )
-    success_json(files: [ new_file.to_hash(@request) ])
+    raise Etna::Error.new("File #{revision.source.path} not found", 404) unless file&.has_data?
```
This method definitely seems like it should be on the revision object. Generally I think it seems preferable to have all of the raising occurring in the controller, and have service objects handle the problem of determining error states. In this case, I think it's especially important not to raise until all of the errors have been collected - since the user may have submitted perhaps hundreds or thousands of revisions, they shouldn't have to discover and repair them one-by-one. The revision object can just collect the errors in an @errors method, which the controller can retrieve and compose an error message containing them. This probably allows the controller to only raise in one place, although it will almost certainly be forced to label everything as "bad request" rather than returning individual statuses.

coleshaw replied:
K, that makes sense. I'll see if I can work that out.

graft commented on lib/server/controllers/metis_controller.rb:
```diff
@@ -29,5 +29,23 @@ def require_folder(bucket, folder_path)
       return folder
     end
+
+    def get_file_obj_from_path(path)
```
Could these move into Metis::Path itself?

coleshaw replied:
So one reason I left those here is that they also implicitly do some validation checks as part of the `require_bucket` call. Let me see how I can keep that behavior, especially for `copy` (I think we explicitly do the checks in `bulk_copy` already).

graft commented on lib/server/controllers/file_controller.rb:
```diff
@@ -72,36 +74,116 @@ def rename
+      project_name: @params[:project_name],
+      owner: ['metis', hmac.id.to_s],
+      name: all_source_bucket_names + all_dest_bucket_names
+    ).all.select{|b| b.allowed?(@user, hmac)}.
```
could be: `name: revision.map(&:bucket_names).flatten`

graft commented on lib/server/controllers/file_controller.rb:
```diff
@@ -72,36 +74,116 @@ def rename
+    end
+
+    # If we've gotten here, every revision looks good and we can execute them!
+    new_files = []
```
You might consider moving this operation to the revision object also, in which case you could do: `success_json(files: revisions.map(&:revise!))`

coleshaw replied:
Good idea!

graft commented:
Very nice work. I think you can finish it up here, it's not that big, yet, and it seems nearly done(:crossed_fingers:). I think you should basically push more behavior out of the controller into the revision object (basically all of the validation stuff that is still in the controller); the controller I think will end up nice and lean, basically just laying out the high-level logic and raising errors, with all of the work happening in the revision object.

coleshaw commented:
@graft , I think I've addressed all your comments about migrating validation and revision execution to `Revision` and keeping the Controller lean. Did I miss anything?

graft commented on lib/path.rb:
```diff
@@ -0,0 +1,76 @@
+          (?<project_name>(#{FILENAME_MATCH.source}))
+          /
+          (?<bucket_name>(#{FILENAME_MATCH.source}))
+          /
```
I think both project_name and bucket_name are only \w+; also I worry that FILENAME_MATCH here will diverge from the one in Metis::File. I guess you can't build the constant using Metis::File::FILENAME_MATCH because of load order issues? Alternatively you could build it on-demand in a memoized class method (like `def self.filepath_match; @filepath_match ||= etc.; end`)

coleshaw replied:
That's a great point, especially if we decide to loosen up the restrictions on the filename. Load order does prevent just using `Metis::File::FILENAME_MATCH` directly, so I've changed it to a memoized class method.

coleshaw commented:
@graft , does this help clarify the validations and what is happening?

coleshaw commented:
@graft , sorry it's taken me so long to reach this point! Per our conversation, I've made the following changes:

* Added specs for `bulk_copy` to check for multiple source and destination buckets when using `bulk_copy` (with no errors and with errors)
* Error messages return as JSON instead of string, using `failure()` instead of Etna errors.
* Error messages are clearer, and for `bulk_copy` include the actual revisions associated with each error.

I will write up the analysis / thoughts we had for doing the validation stuff into #74.

This PR depends on mountetna/etna#50 -- could you take a look at that, too? If it looks good, I can merge and would be happy to publish if you set me up on RubyGems.

coleshaw commented:
@graft, looks okay to merge?

graft commented:
So much work! We should merge. One thing I would change is - in the specs many of the comments have diverged from their intent (especially in `file_spec.rb`) - could you amend these, then :shipit: ?
